### PR TITLE
[US2703] support list snapshots capability in zrepl

### PR DESCRIPTION
**Pull request by mynktl**
Changes : 
Adding the handling of snapshot list command in ZREPL.
A response message from ZREPL will contain the snapshot list and properties in JSON format as described below : 
```json
{
    "snapshot”: [ {
        {
            “name” : “snapshot1 name”,
            “properties”  : { 
                “prop1”:”val1”,
                “prop2”: “val2”,
            },
        },
        {
            “name” : “snapshot2 name”,
            “properties”  : { 
                “prop1”:”val1”,
                “prop2”: “val2”,
                “prop3” : “val3”
            },
        },]
}
```
This is an initial version of changes. By default, we are sending all the properties of the snapshot to the controller.

### CODE REVIEW

File: lib/libzrepl/mgmt_conn.c [hunk 1]
```diff
@@ -1017,6 +1159,32 @@ process_message(uzfs_mgmt_conn_t *conn)
 		    payload_size);
 		break;
 
+	case ZVOL_OPCODE_SNAP_LIST:
+		if (payload_size == 0 || payload_size > MAX_NAME_LEN) {
+			rc = reply_nodata(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq);
+			break;
+		}
+		strlcpy(zvol_name, payload, payload_size);
```
Comment (jkryl) [inline lib/libzrepl/mgmt_conn.c]: payload_size -> sizeof (zvol_name)

File: lib/libzrepl/mgmt_conn.c [hunk 2]
```diff
@@ -555,6 +558,145 @@ uzfs_zvol_stats(uzfs_mgmt_conn_t *conn, zvol_io_hdr_t *hdrp, zvol_info_t *zinfo)
...
+
+	bzero(&hdr, sizeof (hdr));
+	hdr.version = REPLICA_VERSION;
+	hdr.opcode = hdrp->opcode;
+	hdr.io_seq = hdrp->io_seq;
+	hdr.len = total_len + sizeof (*snap_list);
+	hdr.status = ZVOL_OP_STATUS_OK;
+
+	return (reply_data(conn, &hdr, snap_list,
+	    total_len + sizeof (*snap_list)));
```
Comment (jkryl) [inline lib/libzrepl/mgmt_conn.c]: if i understand correctly this is called on behalf of asynchronous thread dispatched from mgmt thread. You can't touch tcp connection or send anything on it from this context. Only the mgmt thread can do such things. So once you complete the list of snapshots, you should return this result back to the mgmt thread and call reply_data() there. Just like we do with other async operations.

File: lib/libzrepl/mgmt_conn.c [hunk 3]
```diff
@@ -1017,6 +1159,32 @@ process_message(uzfs_mgmt_conn_t *conn)
...
+		if (zinfo->mgmt_conn != conn) {
+			uzfs_zinfo_drop_refcnt(zinfo);
+			LOGERRCONN(conn, "Target used invalid connection for "
+			    "zvol %s", zvol_name);
+			rc = reply_nodata(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq);
+			break;
+		}
+		LOGCONN(conn, "Snaplist command for %s", zinfo->name);
+		rc = uzfs_zvol_fetch_snapshot_list(conn, hdrp, zinfo);
```
Comment (vishnuitta) [inline lib/libzrepl/mgmt_conn.c]: this need to be called from async thread, as, we have a single mgmt_thread for all volumes. Doing this call here can block mgmt_thread for other volumes as well.

Comment (vishnuitta) [reply lib/libzrepl/mgmt_conn.c]: please look at SNAPCREATE opcode

File: lib/libzrepl/mgmt_conn.c [hunk 4]
```diff
@@ -1017,6 +1159,32 @@ process_message(uzfs_mgmt_conn_t *conn)
 		    payload_size);
 		break;
 
+	case ZVOL_OPCODE_SNAP_LIST:
+		if (payload_size == 0 || payload_size > MAX_NAME_LEN) {
```
Comment (vishnuitta) [inline lib/libzrepl/mgmt_conn.c]: this code seems to be repetition of SNAPCREATE opcode. Need to remove duplicate code.

**Pull request closed without merging**